File: 3291-find-if-array-can-be-sorted/3291-find-if-array-can-be-sorted.py
class Solution:
    def canSortArray(self, nums: List[int]) -> bool:
        # Bubble Sorting
        n = len(nums)
        for i in range(n):
            flag = False

            for j in range(n - i - 1):
                if(nums[j] <= nums[j + 1]):
                    continue
                if(self.setBits(nums[j]) == self.setBits(nums[j + 1])):
                    nums[j], nums[j + 1] = nums[j + 1], nums[j]
                    flag = True
                else:
                    return False
            if not flag:
                break
        return True
    
    def setBits(self, number):
        count = 0
        while(number > 0):
            # if(number % 2 == 1):
            #     count += 1
            count += number & 1
            # number = number // 2
            number >>= 1
        return count

        

File: 3291-find-if-array-can-be-sorted/3291-find-if-array-can-be-sorted.java
class Solution {
    public boolean canSortArray(int[] nums) {
        // TC- O(n)
        // SC- O(1)
        int currentMin = nums[0];
        int currentMax = nums[0];
        int previousMax = Integer.MIN_VALUE;

        for(int num : nums) {
            if(setBits(num) == setBits(currentMin)) {
                currentMin = Math.min(currentMin, num);
                currentMax = Math.max(currentMax, num);
            } else{
                if(currentMin < previousMax) return false;

                previousMax = currentMax;
                currentMin = currentMax = num;
            }
        }
        // for the last case
        if(currentMin < previousMax) return false;
        return true;
    }
    private int setBits(int number){
        int count  = 0;
        while(number > 0){
            count += number & 1;
            number >>= 1;
        }
        return count;
    }
}